src/ast/ast_validations.c

```diff
@@ -538,665 +702,748 @@ static AST_Validation _ValidateInlinedProperties(const cypher_astnode_t *props)
-			_ValidateFunctionCalls(action, false);
-
-			cypher_astnode_type_t type = cypher_astnode_type(action);
-			if(type == CYPHER_AST_ON_CREATE) {
-				// ON CREATE.
-				uint n_set_items = cypher_ast_on_create_nitems(action);
-				for(uint j = 0; j < n_set_items; j ++) {
-					const cypher_astnode_t *set_item = cypher_ast_on_create_get_item(action, j);
-					if(cypher_astnode_type(set_item) == CYPHER_AST_SET_PROPERTY) {
-						res = Validate_SETProperty(set_item);
-						if(res != AST_VALID) goto cleanup;
-					}
-				}
-			} else if(type == CYPHER_AST_ON_MATCH) {
-				// ON MATCH.
-				uint n_set_items = cypher_ast_on_match_nitems(action);
-				for(uint j = 0; j < n_set_items; j ++) {
-					const cypher_astnode_t *set_item = cypher_ast_on_match_get_item(action, j);
-					if(cypher_astnode_type(set_item) == CYPHER_AST_SET_PROPERTY) {
-						res = Validate_SETProperty(set_item);
-						if(res != AST_VALID) goto cleanup;
-					}
-				}
-			} else {
-				ASSERT(false);
+		for(uint j = 0; j < proj_count; j++) {
+			const cypher_astnode_t *proj = cypher_ast_return_get_projection(return_clause, j);
+			const cypher_astnode_t *alias_node = cypher_ast_projection_get_alias(proj);
+			if(alias_node == NULL)  {
+				// The projection was not aliased, so the projection itself must be an identifier.
+				alias_node = cypher_ast_projection_get_expression(proj);
+				ASSERT(cypher_astnode_type(alias_node) == CYPHER_AST_IDENTIFIER);
+			}
+			const char *alias = cypher_ast_identifier_get_name(alias_node);
+			if(strcmp(projections[j], alias) != 0) {
+				ErrorCtx_SetError("All sub queries in a UNION must have the same column names.");
+				res = AST_INVALID;
+				goto cleanup;
 			}
 		}
 	}
 
 cleanup:
-	array_free(merge_clause_indices);
-	if(defined_aliases) raxFree(defined_aliases);
+	array_free(return_indices);
 	return res;
 }
 
-// Validate each entity referenced in a single path of a CREATE clause.
-static AST_Validation _Validate_CREATE_Entities(const cypher_astnode_t *path,
-												rax *defined_aliases) {
-	if(_ValidateInlinedPropertiesOnPath(path) != AST_VALID) return AST_INVALID;
-
-	uint nelems = cypher_ast_pattern_path_nelements(path);
-	 // Redeclaration of a node is not allowed only when the path is of length 0, as in: MATCH (a) CREATE (a).
-	 // Otherwise, using a defined alias of a node is allowed, as in: MATCH (a) CREATE (a)-[:E]->(:B)
-	if(nelems == 1) {
-		const cypher_astnode_t *node = cypher_ast_pattern_path_get_element(path, 0);
-		const cypher_astnode_t *identifier = cypher_ast_node_pattern_get_identifier(node);
-		if(identifier) {
-			const char *alias = cypher_ast_identifier_get_name(identifier);
-			if(raxFind(defined_aliases, (unsigned char *)alias, strlen(alias)) != raxNotFound) {
-				ErrorCtx_SetError("The bound variable '%s' can't be redeclared in a CREATE clause", alias);
-				return AST_INVALID;
-			}
-		}
+// validate a CALL clause
+static bool _Validate_CALL_Clause
```

Comment:
Not sure, but maybe all validations on the `CALL` clause can be performed when `start = true`

Reply:
Actually we can't promote all of them to occur when `start = true`, since we remove the expression identifiers when we finish traversing the CALL clause node. I'll promote the others.